crates/subspace-networking/src/shared.rs

```diff
@@ -79,3 +85,61 @@ impl Shared {
         }
     }
 }
+
+// Supports exact keys for Kademlia. It uses the Multihash as the canonical type
+// and reduces all the conversions to Multihash conversion.
+// It supports only Sr25519 key pairs for PeerId and Identity multihashes.
+#[derive(Clone, Debug)]
+pub struct IdendityHash(GenericArray<u8, U32>);
+
+impl PreimageIntoKeyBytes<Multihash> for IdendityHash {
+    fn preimage_into_key_bytes(multihash: &Multihash) -> KeyBytes<Self> {
+        const SHA256_HASH_SIZE: usize = 32;
+
+        // Identity hasher for 32-bytes digest
+        if multihash.code() == u64::from(Code::Identity)
+            && multihash.digest().len() == SHA256_HASH_SIZE
+        {
+            let array = GenericArray::from_slice(multihash.digest());
+            return KeyBytes::from_unchecked(*array);
+        }
+
+        panic!(
+            "Unsupported multihash type. Expected Identity:{:?}",
+            multihash
+        )
+    }
+}
+
+impl PreimageIntoKeyBytes<PeerId> for IdendityHash {
+    fn preimage_into_key_bytes(peer_id: &PeerId) -> KeyBytes<Self> {
+        let multihash: Multihash = if let Ok(public_key) =
+            identity::PublicKey::from_protobuf_encoding(peer_id.as_ref().digest())
+        {
+            if let identity::PublicKey::Sr25519(pk) = public_key {
+                Code::Identity.digest(&pk.encode())
```

Comment:
This transforms public key into it into `Multihash` just to extract it from `Multihash` using `impl PreimageIntoKeyBytes<Multihash> for IdentityHash`. I feel like it would make more sense to turn it into `KeyBytes` right here.

Reply:
I did it intentionally to have a single conversion algorithm. I'd rather leave it as it is because IMHO it's easier to debug and enhance the code this way.

Reply:
It is subjective for sure, but to me the less `impl`s I need to read through to understand how `PeerId` is turned into bytes the better.

Reply:
The original `Key` type in the Kademlia's master branch contains multiple copies of the conversion code: https://github.com/libp2p/rust-libp2p/blob/ea487aebfe6eb672b05d2bec2d9d79bbd92450ba/protocols/kad/src/kbucket/key.rs#L97

When we need to change it we must change in multiple places.